python/scope.py

```diff
@@ -506,6 +508,15 @@ def type_check_expr(self, expr):
 
 		elif expr.data == "tupleval":
 			return [self.type_check_expr(e) for e in expr.children]
+		elif expr.data == "listval":
+			if (len(expr.children) == 0):
+				return [lark.Token("LIST", "list")]
+
+			for i,e in enumerate(expr.children):
+				if self.type_check_expr(e) != self.type_check_expr(expr.children[0]):
```

Comment:
I think calling `type_check_expr` multiple times on the same child will result in duplicate warnings. You probably should cache the types and then enumerate over them, like

```py
first, *rest = [self.type_check_expr(e) for e in expr.children]

for i, e in enumerate(rest):
```